Source/drlg_l2.cpp

```diff
@@ -2880,35 +2905,8 @@ void GenerateLevel(lvl_entry entry)
 		}
 		FloodTransparencyValues(3);
 		FixTransparency();
-		if (entry == ENTRY_MAIN) {
-			doneflag = PlaceMiniSet(USTAIRS, 1, 1, -1, -1, true);
-			if (doneflag) {
-				doneflag = PlaceMiniSet(DSTAIRS, 1, 1, -1, -1, false);
-				if (doneflag && currlevel == 5) {
-					doneflag = PlaceMiniSet(WARPSTAIRS, 1, 1, -1, -1, false);
-				}
-			}
-			ViewPosition.y -= 2;
-		} else if (entry == ENTRY_PREV) {
-			doneflag = PlaceMiniSet(USTAIRS, 1, 1, -1, -1, false);
-			if (doneflag) {
-				doneflag = PlaceMiniSet(DSTAIRS, 1, 1, -1, -1, true);
-				if (doneflag && currlevel == 5) {
-					doneflag = PlaceMiniSet(WARPSTAIRS, 1, 1, -1, -1, false);
-				}
-			}
-			ViewPosition.x--;
-		} else {
-			doneflag = PlaceMiniSet(USTAIRS, 1, 1, -1, -1, false);
-			if (doneflag) {
-				doneflag = PlaceMiniSet(DSTAIRS, 1, 1, -1, -1, false);
-				if (doneflag && currlevel == 5) {
-					doneflag = PlaceMiniSet(WARPSTAIRS, 1, 1, -1, -1, true);
-				}
-			}
-			ViewPosition.y -= 2;
-		}
-	}
+		doneflag = PlaceStairs(entry);
+	} while (!doneflag);

```

Comment:
You could even do this as `while (true) { if (PlaceStairs()) { break; } }` :D

would remove the need for `doneFlag`

Reply:
why not for(;;)
:D

Reply:
@ephphatha done, I also flattened the one for caves.